src/compiler/emitter.ts

```diff
@@ -2591,57 +2737,117 @@ namespace ts {
             }
 
             if (format & ListFormat.BracketsMask) {
-                write(getClosingBracket(format));
+                writePunctuation(getClosingBracket(format));
+            }
+        }
+
+        function commitPendingSemicolon() {
+            if (pendingSemicolon) {
+                writer.writePunctuation(";");
             }
+            pendingSemicolon = false;
+        }
+
+        function writeLiteral(s: string) {
+            commitPendingSemicolon();
+            writer.writeLiteral(s);
+        }
+
+        function writeStringLiteral(s: string) {
+            commitPendingSemicolon();
+            writer.writeStringLiteral(s);
         }
 
-        function write(s: string) {
+        function writeBase(s: string) {
+            commitPendingSemicolon();
             writer.write(s);
         }
 
+        function writeSymbol(s: string, sym: Symbol) {
+            commitPendingSemicolon();
+            writer.writeSymbol(s, sym);
+        }
+
+        function writePunctuation(s: string) {
+            commitPendingSemicolon();
+            if (printerOptions.omitTrailingSemicolon && s === ";") {

```

Comment:
Add a `writeSemicolon()` function for this case (and replace every `writePunctuation(";")`) so we don't have to observe the cost of the lookup and string comparison for every punctuation mark.